### PR TITLE
create nextEpochMap type

### PR DESCRIPTION
**Pull request by timwu20**
## Changes

- `nextEpochMap` type which uses generics

## Tests

```sh
go test -tags integration github.com/ChainSafe/gossamer
```

## Primary Reviewer

@EclesioMeloJunior

### CODE REVIEW

**Comment by EclesioMeloJunior** [comment]
Cool stuff 👍 ! Feel free to merge your modifications!

File: dot/state/epoch.go [hunk 1]
```diff
@@ -349,7 +349,8 @@ func (s *EpochState) GetConfigData(epoch uint64, header *types.Header) (configDa
 		// we will check in the memory map and if we don't find the data
 		// then we continue searching through the previous epoch
 		s.nextConfigDataLock.RLock()
-		inMemoryConfigData, err := retrieveFromMemory(s.nextConfigData, s.blockState, uint64(tryEpoch), header)
+		inMemoryConfigData, err := nextEpochMap[types.NextConfigData](s.nextConfigData).Retrieve(s.blockState, uint64(tryEpoch), header)
+		// inMemoryConfigData, err := retrieveFromMemory(s.nextConfigData, s.blockState, uint64(tryEpoch), header)
```
Comment (EclesioMeloJunior) [inline dot/state/epoch.go:353]:
```suggestion
		inMemoryConfigData, err := s.nextConfigData.Retrieve(s.blockState, uint64(tryEpoch), header)
```

File: dot/state/epoch.go [hunk 2]
```diff
@@ -380,10 +381,10 @@ func (s *EpochState) getConfigDataFromDatabase(epoch uint64) (*types.ConfigData,
 	return info, nil
 }
 
-func retrieveFromMemory[T types.NextEpochData | types.NextConfigData](nextEpochMap map[uint64]map[common.Hash]T,
-	blockState *BlockState, epoch uint64, header *types.Header) (*T, error) {
+type nextEpochMap[T types.NextEpochData | types.NextConfigData] map[uint64]map[common.Hash]T
 
-	atEpoch, has := nextEpochMap[epoch]
+func (nem nextEpochMap[T]) Retrieve(blockState *BlockState, epoch uint64, header *types.Header) (*T, error) {
```
Comment (EclesioMeloJunior) [inline dot/state/epoch.go:386]: could it be not exported?

**Review by EclesioMeloJunior — CHANGES_REQUESTED** [review]

File: dot/state/epoch.go (hunk 2 above)
Comment (timwu20) [reply dot/state/epoch.go:385]: well the type is private, but it seems like there should be an exported method on any type IMO.

**Review by EclesioMeloJunior — APPROVED** [review]

**Pull request merged**